oauth2_client/client.py

```diff
@@ -45,7 +88,21 @@ def request(self, method, url, *args, **kwargs):  # pylint: disable=arguments-di
             Response object.
         """
         absolute_url = urljoin(self.service_host, url)
-        return super(OAuth2Client, self).request(method, absolute_url, *args, **kwargs)
+        try:
+            resp = super(OAuth2Client, self).request(method, absolute_url, *args, **kwargs)
+            if resp.status_code == 403:
+                raise TokenExpiredError()
+            else:
+                return resp
+        except TokenExpiredError:
+            log.warning("Access Token for app=%s expired.", self.app.name)
+            if self.expired_callback:
+                new_token = self.expired_callback(self.app)
+                self.token = new_token.to_client_dict()
+                log.info("Obtained new token. Retrying request.")
+                self.request(method, url, *args, **kwargs)
+            else:
+                raise
```

Comment:
I'd personally prefer to reduce nesting levels and rewrite the code this way:
```py
        resp = super(OAuth2Client, self).request(method, absolute_url, *args, **kwargs)
        if resp.status_code != 403:
            return resp
            
        log.warning("Access Token for app=%s expired.", self.app.name)
        
		if not self.expired_callback:
            raise TokenExpiredError()
            
        new_token = self.expired_callback(self.app)
        self.token = new_token.to_client_dict()
        log.info("Obtained new token. Retrying request.")
        return self.request(method, url, *args, **kwargs)
```
as it increases code readability.
see https://en.wikibooks.org/wiki/Computer_Programming/Coding_Style/Minimize_nesting

Reply:
The try/except has to stay there, as the call to `super.request...` can internally raise a `TokenExpiredError` too. However I managed to reduce the nesting a little.

Although personally I think I am rather a fan of "balanced nesting" rather than “Flat is better than nested”.

Reply:
Correct, I missed `super.request` part.
Another one question:

Isn't `super.request` already handle automatic token refreshing? Does it make sense to re-use it instead of introducing custom implementation?
https://github.com/requests/requests-oauthlib/blob/master/requests_oauthlib/oauth2_session.py#L65-L74

Reply:
only with flows that use refresh token:
https://github.com/requests/requests-oauthlib/blob/792f1849089f87da59cb29118053f6d6f1bd199c/requests_oauthlib/oauth2_session.py#L377 

So won't work with e.g. JWT flow

Reply:
@fredmajor could extend the docstring with the information about token refreshing flow?
BTW, do we need to catch `TokenExpiredError` and try to get a new toke for non-JWT flows, as it could be already done by `super.request`?
Does it make sense to move GrantType specific behavior to a separate class?

Reply:
I investigated that again to be sure. 
Seems that we won't be able to use the built-in token refresh based on the refresh_token. Why? Because both flows we are to currently support don't provide that token :D 

> Client credentials grant: A refresh token SHOULD NOT be included. (RFC)
JWT flow: NOTE This flow never issues a refresh token.  

Our in-house provider seems to not return the `refresh_token`, I checked. I also self-thoroughly skimmed through the django-oauth-toolkit docs and there is no obvious way to make it return so. They chose to follow the rfc.

I will put that info in the docs and leave the current solution in the code as it should do the job.